### PR TITLE
Alert user upon switching to unconnected account

### PR DESCRIPTION
**Pull request by Gudahtt**
An alert is now shown when the user switches from an account that is connected to the active tab to an account that is not connected. The alert prompts the user to dismiss the alert or connect the account they're switching to.

The "loading" state is handled by disabling the buttons, and the error state is handled by displaying a generic error message and disabling the connect button.

The new reducer for this alert has been created with `createSlice` from the Redux Toolkit. This utility is recommended by the Redux team, and represents a new style of writing reducers that I hope we will use more in the future (or at least something similar). `createSlice` constructs a reducer, actions, and action creators automatically. The reducer is constructed using their `createReducer` helper, which uses Immer to allow directly mutating the state in the reducer but exposing these changes as immutable.

### CODE REVIEW

**Comment by Gudahtt** [comment]
I'll wait on #8345 before fixing the broken integration tests, as I'll have one less state file to modify after that's merged.

**Comment by Gudahtt** [comment]
<details>
<summary>Screenshots:</summary>

Here's the alert:

normal

Here is the error state:

error

</details>

File: ui/app/ducks/alerts/unconnected-account.js
Comment (Gudahtt) [inline ui/app/ducks/alerts/unconnected-account.js:51]: Nit: typo, this should say `close`, not `lose`
```suggestion
      // close the alert if the account is switched while it's open
```

File: ui/app/components/app/alerts/alerts.scss
Comment (Gudahtt) [inline ui/app/components/app/alerts/alerts.scss:1]: Nit: missing semi-colon at the end of this line

```suggestion
@import './unconnected-account-alert/unconnected-account-alert.scss';
```

File: ui/app/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js
Comment (Gudahtt) [inline ui/app/components/app/alerts/unconnected-account-alert/unconnected-account-alert.js:60]: Nit: This should be a self-closing, since there is nothing being passed in as children.
```suggestion
    />
```

File: ui/app/components/app/alerts/alerts.js
Comment (Gudahtt) [inline ui/app/components/app/alerts/alerts.js:8]: Nit: this is a poor choice in variable name, since I'll be adding similar variables for each other alert as well (they can't all be called `isOpen`).

File: ui/app/ducks/alerts/unconnected-account.js
Comment (Gudahtt) [inline ui/app/ducks/alerts/unconnected-account.js:49]: This doesn't need to block review, but I'm working on an improvement to this reducer that would remove the necessity to cache the current selected address. This approach of caching the relevant background state and watching for state updates doesn't scale well when you want to watch a larger set of things, and keeping redundant state in reducers is generally not advised. This was particularly noticeable while I was working on the next alert, which also needs to watch for changes to the current active tab.

**Review by rekmarks — CHANGES_REQUESTED** [review]
Very clean and crisp! A pleasure to review. Excited to see Immer being used, even if indirectly.

One in-line comment, and also some feedback about styling/UX:

image

<details>
<summary><strong>Full Screenshot for Scale</strong></summary>

</details>

- The buttons appear disproportionately large to me, relative to the size of the popover. Could we make them smaller?
  - Related: Where are the designs?
- In testing this, I quickly got very annoyed with this popup. Can we add a "Don't show this again" checkbox?

File: test/unit/ui/app/actions.spec.js
Comment (Gudahtt) [reply test/unit/ui/app/actions.spec.js:334]: This was added so that these [two selectors would ](https://github.com/MetaMask/metamask-extension/pull/8312/files#diff-eebb93d0666faf7aa39f85d9f7efa804R1204) not blow up. I should replace this with a more minimal state.

Comment (rekmarks) [reply test/unit/ui/app/actions.spec.js:334]: Got it. Definitely won't block on that, whatever you do!

**Comment by Gudahtt** [comment]
Yeah, the buttons do look a bit large. I think those sizes are being dictated by our default button styles? I'll take a look. We have another card for updating the button styles across the extension to be rounded, as shown in the Figma designs, but that wasn't going to be in v8. Maybe I can tweak them to be a bit smaller though.

The designs are [here](https://www.figma.com/file/CvD7nhQVeSEB1QuSy3XhtY/oCap-%2F-Login-Per-Site?node-id=1929%3A139).

The "Don't show this again" checkbox was going to be in a follow-up PR; there's a separate card for that on Trello I believe.

**Comment by rekmarks** [comment]
@Gudahtt word, sounds good! Just me when you've decided what to do with the button styles.

Oh one more thing: doing any e2e tests for this in a follow-up, or what's the plan for that?

File: test/unit/ui/app/actions.spec.js
Comment (Gudahtt) [reply test/unit/ui/app/actions.spec.js]: I've replaced these with a more minimal state (just the selected address).

**Review by rekmarks — DISMISSED** [review]
LGTM!

Buttons and checkbox can be fixed in one or more follow-up PRs.

File: ui/app/ducks/alerts/unconnected-account.js
Comment (Gudahtt) [reply ui/app/ducks/alerts/unconnected-account.js:45]: This has been completed in bf55bbaee0f75d0230a16525cc6073a887dbc65b

**Comment by Gudahtt** [comment]
Two updates have been made:

* Use `SELECTED_ADDRESS_CHANGED` state instead of cached address

  The `unconnected-account` reducer no longer caches the selected address when the alert was opened. Instead a new event was added for when the selected address changes (`SELECTED_ADDRESS_CHANGED`). Now that we know exactly when the selected address changes, we don't need to cache the original account for comparison anymore.

* Update button styles to reduce size

  The buttons are now closer in size to the buttons in the designs. The button styles will be updated to match the designs more closely in a separate PR.

<details>
<summary>Screenshots of updated buttons:</summary>

updated-buttons

</details>

**Comment by Gudahtt** [comment]
@rekmarks I started trying to write an e2e test for this, but it was a bit challenging because we can't actually test the real popup with Selenium. It doesn't allow interacting with browser-things outside of the page (even the keyboard shortcut is intercepted). Even if it did allow it, there's no way to focus on the popup either.

In other tests we fake the popup UI by navigating to `popup.html` and making the window vagely popup-sized. The `activeTab` state still won't be set in that case though, and I rely upon that for this alert. We'll need to find a way to fake the `activeTab` state - then this should be testable.

**Review by rekmarks — DISMISSED** [review]
LGTM!

Nice fix for the buttons and selected address state.

**Review by rekmarks — APPROVED** [review]
Re-approve after rebase.

**Review by whymarrh — APPROVED** [review]
This looks good to me! I didn't test it out locally but it looks like @rekmarks did.

**Pull request merged**
